### PR TITLE
70 update the tests in tests/test_github_repo_request_local.py

### PR DESCRIPTION
**Pull request by tnzmnjm**
Hi Julian,

Could you please review the changes and advise if there's any section you would like me to change please?

This PR introduces several updates to our test suites to improve accuracy and robustness, and it proposes a change in file naming to better reflect the content and purpose of our scripts.

Changes:

1. Update Test Cases for `extract_and_flag_domains`:
- Modified the test cases to handle `pd.NA` correctly, ensuring that our tests remain robust across mixed data types.

2. Correct Test Assertions in `mark_incomplete_urls`:
- Aligned the test assertions with the function's logic to accurately reflect the criteria for URL completeness, especially handling special characters within URLs.

3. Adjust Expected Outputs in Tests:
- Updated the expected outputs in our tests to match actual behaviour, focusing particularly on the handling of unsupported URL schemes and duplicate flags.

4. Improve Test Coverage:
Added new scenarios to our tests to thoroughly examine the interactions with the `duplicate_flag` and `URL schemes`. 

5. File Name Change:
- Propose renaming of `tests/test_github_repo_request_local.py` to `tests/test_initial_data_preparation.py` as the test cases are written for the functions in the script `initial_data_preparation.py`.

6. Wrote unit tests using `PyHamcrest` for the functions `get_domain`, `is_complete_url`, `extract_url`, and `convert_http_to_https` 

Thanks very much

### CODE REVIEW

File: tests/test_initial_data_preparation.py
Comment (julianharty) [inline tests/test_initial_data_preparation.py:1]: Please add a description of the purpose of this file to the top of the file.

File: tests/test_initial_data_preparation.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+                "https://hydrillabugs.koszko.org/projects/haketilo/repository",
+                "https://git.replicant.us/replicant",
+                "https://gerrit.osmocom.org/plugins/gitiles/ims",
+            ],
+            "duplicate_flag": [False] * 23,
+            "unsupported_url_scheme": [False] * 23,
+            "incomplete_url_flag": [False] * 23,
+        }
+    )
+    data.loc[10:11, "unsupported_url_scheme"] = True  # Invalid URL formats
```
Comment (julianharty) [inline tests/test_initial_data_preparation.py:45]: I realise this is effectively copied from the previous code, nonetheless how about adding a brief note to say that 10 and 11 align with the entries?: 
```
                "just-a-string",
                "http:///a-bad-url.com",
```

File: tests/test_initial_data_preparation.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+def test_get_base_repo_url():
+    # Test data setup
+    data = pd.DataFrame(
+        {
+            "repourl": [
+                None,  # Test for None input
+                "https://github.com/getdnsapi/stubby",
+                "https://github.com/namecoin",
+                "https://github.com/osresearch/heads/issues/540",
```
Comment (julianharty) [inline tests/test_initial_data_preparation.py:18]: Can you add a valid url that starts with http:// please?

File: tests/test_initial_data_preparation.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+                    False,
+                    False,
+                    False,
+                    False,
+                    False,
+                    False,
+                    False,
+                ],
+            },
+            [False, True, True, True, True, True, True, True],
```
Comment (julianharty) [inline tests/test_initial_data_preparation.py:140]: It's not obvious to me what these False, True... mean and I can't easily reconcile them with lines 108 to 116. Perhaps you could add a note to explain the mapping of these values?

File: tests/test_initial_data_preparation.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        mark_incomplete_urls(df)
+
+    # Check that the error message is as expected
+    assert (
+        str(exc_info.value) == "DataFrame must contain a 'repourl' column."
+    ), "Error message does not match the expected output."
+
+
+@pytest.mark.parametrize(
+    "data, expected_flags",
```
Comment (julianharty) [inline tests/test_initial_data_preparation.py:103]: what values can the expected_flags (note this is plural not singular) contain and what do they represent? FWIW I'd expect an `expected_flag` (singular) to be able to contain a single item - here it seems to be `True` or `False` while I'd expect an `expected_flags` to support more than one item.

File: tests/test_initial_data_preparation.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+            ],
+            [False, False, False, True, True],
+        ),
+        # Test with a mix of duplicate and non-duplicate entries
+        (
+            {
+                "repourl": [
+                    "https://github.com/ownwe/openai",
+                    # ftp is not a list of accepted schemes ["http", "https",
+                    # "git"]
```
Comment (julianharty) [inline tests/test_initial_data_preparation.py:249]: `..."git" are the accepted schemes.`

File: tests/test_helper_functions.py
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply tests/test_helper_functions.py:1]: It has been updated. I pushed the code after finished writing the unit tests.

File: tests/test_initial_data_preparation.py
Comment (tnzmnjm) [reply tests/test_initial_data_preparation.py:1]: - Added `docstring`s for the script and its functions.

File: tests/test_initial_data_preparation.py (hunk 1 above)
Comment (tnzmnjm) [reply tests/test_initial_data_preparation.py:64]: This is done.

File: tests/test_initial_data_preparation.py (hunk 2 above)
Comment (tnzmnjm) [reply tests/test_initial_data_preparation.py:37]: I have added `http://git.savannah.gnu.org/cgit/gnucap.git` --> The test passes. 
However, I do not expect to see any `repourl` starting with `http` here as `http` is converted to 'https' in the lines before the call for the function `get_base_repo_url` .

File: tests/test_initial_data_preparation.py (hunk 4 above)
Comment (tnzmnjm) [reply tests/test_initial_data_preparation.py:130]: `expected_flags` is a list of boolean values that represent if their respective item in the list of urls are expected to be flagged as incomplete(True) or not(False).

I understand as each url will be flagged with only one of two values, using a singular form is appropriate.

File: tests/test_initial_data_preparation.py (hunk 5 above)
Comment (tnzmnjm) [reply tests/test_initial_data_preparation.py]: This is fixed.

File: tests/test_helper_functions.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        # Test with an empty string
+        assert_that(is_complete_url(""), is_(False))
+
+
+# Tests for the function `extract_url`
+class TestExtractURL(unittest.TestCase):
+    def test_empty_input(self):
+        result, flag = extract_url("")
+        assert_that(result, is_(None))
+        assert_that(flag, is_(True))
```
Comment (julianharty) [inline tests/test_helper_functions.py:87]: The variable name `flag` doesn't communicate the purpose of the flag, so when reading the various `assert_that` statements here and elsewhere in these tests that test the value of `flag` don't read well. Please can you find a more descriptive name to replace `flag`?

File: tests/test_initial_data_preparation.py (hunk 2 above)
Comment (julianharty) [reply tests/test_initial_data_preparation.py:37]: IIRC there are 15 URLs in the dataset from NLnet that have `http://` e.g. `http://devel.ag-projects.com/repositories/sylkserver`

File: tests/test_initial_data_preparation.py (hunk 3 above)
Comment (julianharty) [reply tests/test_initial_data_preparation.py:181]: Thank you, please add a short comment for this array of values to say the flags represent incomplete URLs.

File: tests/test_initial_data_preparation.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+                    "ftp://example.net/resource",
+                    "smb://fileserver/home",
+                ],
+                "duplicate_flag": [False, False, False, False, False],
+            },
+            [
+                pd.NA if domain is None else domain
+                for domain in ["github.com", "example.com", "example.org", None, None]
+            ],
+            [False, False, False, True, True],
```
Comment (julianharty) [inline tests/test_initial_data_preparation.py:285]: Again please could you add a short comment here to explain what this array contains (in other words its purpose).

**Review by julianharty — APPROVED** [review]
Approved subject to the changes I've requested; thank you very much.

File: tests/test_initial_data_preparation.py (hunk 2 above)
Comment (tnzmnjm) [reply tests/test_initial_data_preparation.py:37]: Yes. I have noticed them as well.

File: tests/test_initial_data_preparation.py (hunk 3 above)
Comment (tnzmnjm) [reply tests/test_initial_data_preparation.py:182]: This is done.

File: tests/test_helper_functions.py (hunk 6 above)
Comment (tnzmnjm) [reply tests/test_helper_functions.py]: Change the variable name `flag` to `extraction_failed'.

**Comment by tnzmnjm** [comment]
All the required changes have been done. Merging this PR to the main branch.

**Review by julianharty — COMMENTED** [review]
Thanks for these changes.

**Pull request merged**